src/test/java/net/minecraftforge/debug/world/MultiNoiseBiomeSourceLoadingEventTest.java

```diff
@@ -0,0 +1,259 @@
+    
+    public MultiNoiseBiomeSourceLoadingEventTest()
+    {
+        IEventBus modBus = FMLJavaModLoadingContext.get().getModEventBus();
+        IEventBus forgeBus = MinecraftForge.EVENT_BUS;
+
+        modBus.addListener(this::onGatherData);
+        forgeBus.addListener(this::onLoadBiomeSource);
+        forgeBus.addListener(this::onRegisterCommands);
+    }
+    
+    private void onGatherData(GatherDataEvent event)
+    {
+        // need dynamic registries to do worldgen datagen
+        RegistryHolder registries = RegistryAccess.builtin();
+        RegistryWriteOps<JsonElement> writeOps = RegistryWriteOps.create(JsonOps.INSTANCE, registries);
+        
+        // the objects we reference must have been registered to the registryaccess
+        // creating a registry access deep-copies most of the BuiltinRegistries objects,
+        // so we must get vanilla objects from the registryaccess instead of getting static references
+        Registry<DimensionType> dimensionTypes = registries.registryOrThrow(Registry.DIMENSION_TYPE_REGISTRY);
+        Registry<NoiseGeneratorSettings> noiseSettings = registries.registryOrThrow(Registry.NOISE_GENERATOR_SETTINGS_REGISTRY);
+        Registry<NoiseParameters> noiseParameters = registries.registryOrThrow(Registry.NOISE_REGISTRY);
+        Registry<Biome> biomes = registries.registryOrThrow(Registry.BIOME_REGISTRY);
+        
+        DimensionType overworldDimensionType = dimensionTypes.get(DimensionType.OVERWORLD_LOCATION);
+        NoiseGeneratorSettings overworldNoiseSettings = noiseSettings.get(NoiseGeneratorSettings.OVERWORLD);
+        Biome desert = biomes.get(Biomes.DESERT);
+        
+        // make two three dimensions for testing
+        // one with a preset biome source, one with a fully custom biome source and a dimension name, one with a fully custom source without name
+        
+        // preset biome source dimension
+        BiomeSource presetSource = MultiNoiseBiomeSource.Preset.OVERWORLD.biomeSource(biomes);
+        ChunkGenerator presetGenerator = new NoiseBasedChunkGenerator(noiseParameters, presetSource, 0, () -> overworldNoiseSettings);
+        LevelStem presetDimension = new LevelStem(() -> overworldDimensionType, presetGenerator);
+        
+        // fully custom biome source dimension with forge:name field
+        List<Pair<ParameterPoint, Supplier<Biome>>> parameterPoints = new ArrayList<>();
+        parameterPoints.add(Pair.of(new ParameterPoint(
+            Parameter.span(0.2F, 0.55F), // temperature
+            Parameter.span(0.3F, 1.0F), // humidity
+            Parameter.span(0.3F, 1.0F), // continentalness
+            Parameter.span(-0.375F, -0.2225F), // erosion
+            Parameter.point(1.0F), // depth
+            Parameter.span(0.9333F, 1.0F), // weirdness
+            0L // offset
+            ), () -> desert));
+        ParameterList<Supplier<Biome>> biomeParams = new ParameterList<>(parameterPoints);
```

Comment:
For readability, this should be moved to above the comment, since this is used by both named and nameless biome sources.

Reply:
Rearranged in next push